--- contrib/ipfilter/ipf.c.orig	Sun Jul  4 11:24:39 2004
+++ contrib/ipfilter/ipf.c	Sun Apr 24 05:37:52 2005
@@ -380,13 +380,11 @@
 				if (ioctl(fd, del, &fr) == -1) {
 					fprintf(stderr, "%d:", linenum);
 					perror("ioctl(delete rule)");
-					exit(1);
 				}
 			} else if (!(opts & OPT_DONOTHING)) {
 				if (ioctl(fd, add, &fr) == -1) {
 					fprintf(stderr, "%d:", linenum);
 					perror("ioctl(add/insert rule)");
-					exit(1);
 				}
 			}
 		}
