--- Makefile.orig	2011-07-11 22:41:12.000000000 +0000
+++ Makefile	2011-07-11 22:57:12.000000000 +0000
@@ -17,7 +17,6 @@
 SRCS+=	af_link.c		# LLC support
 SRCS+=	af_inet.c		# IPv4 support
 SRCS+=	af_inet6.c		# IPv6 support
-SRCS+=	af_atalk.c		# AppleTalk support
 SRCS+=	af_nd6.c		# ND6 support
 
 SRCS+=	ifclone.c		# clone device support
@@ -33,20 +32,14 @@
 
 SRCS+=	ifcarp.c		# SIOC[GS]VH support
 SRCS+=	ifgroup.c		# ...
-SRCS+=	ifpfsync.c		# pfsync(4) support
 
 SRCS+=	ifbridge.c		# bridge support
 SRCS+=	iflagg.c		# lagg support
 
-.if ${MK_IPX_SUPPORT} != "no" && !defined(RELEASE_CRUNCH)
-SRCS+=	af_ipx.c		# IPX support
-DPADD+=	${LIBIPX}
-LDADD+=	-lipx
-.endif
 
 MAN=	ifconfig.8
 
-CFLAGS+= -Wall -Wmissing-prototypes -Wcast-qual -Wwrite-strings -Wnested-externs
+CFLAGS+= -I/usr/m0n0wall/build82/tmp/ -Wall -Wmissing-prototypes -Wcast-qual -Wwrite-strings -Wnested-externs
 WARNS?=	0
 
 .include <bsd.prog.mk>
