--- pfkey.c.orig        2011-05-14 00:53:52.000000000 +0000
+++ pfkey.c     2011-05-14 00:54:35.000000000 +0000
@@ -1776,6 +1776,11 @@
                case SADB_EXT_SPIRANGE:
                case SADB_X_EXT_POLICY:
                case SADB_X_EXT_SA2:
+               case SADB_X_EXT_NAT_T_TYPE:
+               case SADB_X_EXT_NAT_T_SPORT:
+               case SADB_X_EXT_NAT_T_DPORT:
+               case SADB_X_EXT_NAT_T_OAI:
+               case SADB_X_EXT_NAT_T_OAR:
                        mhp[ext->sadb_ext_type] = (caddr_t)ext;
                        break;
                default:
