--- syslogd.c.orig	2008-10-02 03:57:24.000000000 +0100
+++ syslogd.c	2010-09-13 15:43:42.000000000 +0100
@@ -1926,9 +1926,16 @@
 			tp = f->f_un.f_forw.f_hname;
 			p++;
 
-			while (*p && (*p != ':') && (i-- > 0)) {
+			while (*p && (*p != ':') && (*p != '[') && (i-- > 0)) {
 				*tp++ = *p++;
 			}
+			if (*p == '[') {
+				p++;
+				while (*p && (*p != ']') && (i-- > 0)) {
+					*tp++ = *p++;
+				}
+				p++;
+			}
 			*tp = '\0';
 		}
 		/* See if we copied a domain and have a port */
