--- M0N0WALL_GENERIC.orig	2011-05-07 19:50:50.000000000 +0000
+++ M0N0WALL_GENERIC	2011-05-07 21:03:53.000000000 +0000
@@ -1,10 +1,7 @@
-machine		i386
-cpu		I486_CPU
-cpu		I586_CPU
-cpu		I686_CPU
-ident		M0N0WALL_GENERIC
+cpu		HAMMER	
+ident		M0N0WALL_GENERIC64
 
-makeoptions	MODULES_OVERRIDE="dummynet ipfw acpi if_tap if_vlan"
+makeoptions	MODULES_OVERRIDE="dummynet ipfw if_tap if_vlan"
 
 options		HZ=1000
 
@@ -28,16 +25,12 @@
 options 	KBD_INSTALL_CDEV	# install a CDEV entry in /dev
 options 	FLOWTABLE		# per-cpu routing cache
 
-# To make an SMP kernel, the next two lines are needed
+# To make an SMP kernel, the next line is needed
 options 	SMP			# Symmetric MultiProcessor Kernel
-device		apic			# I/O APIC
 
 # CPU frequency control
 device		cpufreq
 
-options		CPU_SOEKRIS
-options		CPU_GEODE
-
 options		NO_SWAPPING
 options         DEVICE_POLLING
 
@@ -54,8 +47,8 @@
 options		IPSTEALTH
 
 # Bus support.
-device		eisa
 device		pci
+device		acpi
 
 # Floppy drives
 device		fdc
@@ -69,7 +62,6 @@
 options 	ATA_STATIC_ID	# Static device numbering
 
 # SCSI Controllers
-device		ahb		# EISA AHA1742 family
 device		ahc		# AHA2940 and onboard AIC7xxx devices
 options 	AHC_REG_PRETTY_PRINT	# Print register bitfields in debug
 					# output.  Adds ~128k to driver.
@@ -87,7 +79,6 @@
 
 device		adv		# Advansys SCSI adapters
 device		adw		# Advansys wide SCSI adapters
-device		aha		# Adaptec 154x SCSI adapters
 device		aic		# Adaptec 15[012]x SCSI adapters, AIC-6[23]60.
 device		bt		# Buslogic/Mylex MultiMaster SCSI adapters
 
@@ -107,7 +98,6 @@
 # RAID controllers interfaced to the SCSI subsystem
 device		amr		# AMI MegaRAID
 device		arcmsr		# Areca SATA II RAID
-device		asr		# DPT SmartRAID V, VI and Adaptec SCSI RAID
 device		ciss		# Compaq Smart RAID 5*
 device		dpt		# DPT Smartcache III, IV - See NOTES for options
 device		hptmv		# Highpoint RocketRAID 182x
@@ -120,10 +110,8 @@
 # RAID controllers
 device		aac		# Adaptec FSA RAID
 device		aacp		# SCSI passthrough for aac (requires CAM)
-device		ida		# Compaq Smart RAID
 device		mfi		# LSI MegaRAID SAS
 device		mlx		# Mylex DAC960 family
-device		pst		# Promise Supertrak SX6000
 device		twe		# 3ware ATA RAID
 
 # atkbdc0 controls both the keyboard and the PS/2 mouse
@@ -137,9 +125,6 @@
 # syscons is the default console driver, resembling an SCO console
 device		sc
 
-# Add suspend/resume support for the i8254.
-device		pmtimer
-
 # PCCARD (PCMCIA) support
 # PCMCIA and cardbus bridge support
 device		cbb		# cardbus (yenta) bridge
@@ -201,7 +187,6 @@
 device		ex		# Intel EtherExpress Pro/10 and Pro/10+
 device		ep		# Etherlink III based cards
 device		fe		# Fujitsu MB8696x based cards
-device		ie		# EtherExpress 8/16, 3C507, StarLAN 10 etc.
 device		sn		# SMC's 9000 series of Ethernet chips
 device		xe		# Xircom pccard Ethernet
 
@@ -259,7 +244,6 @@
 device		cryptodev
 device		hifn
 device		padlock
-device		glxsb
 
 # USB support
 device		uhci		# UHCI PCI->USB interface
