--- sys/contrib/ipfilter/netinet/ip_nat.c.orig	2008-10-02 04:57:24.000000000 +0200
+++ sys/contrib/ipfilter/netinet/ip_nat.c	2009-08-04 23:00:53.000000000 +0200
@@ -2279,7 +2289,7 @@
 			     (u_int)fin->fin_p, inb, fin->fin_src);
 	fin->fin_data[0] = sp;
 	fin->fin_data[1] = dp;
-	if (natl != NULL)
+	if (natl != NULL && natl->nat_redir == NAT_REDIRECT)
 		return -1;
 
 	nat->nat_inip.s_addr = htonl(in.s_addr);
