--- sys/sys/mbuf.h.orig	2008-02-17 21:32:18.000000000 +0100
+++ sys/sys/mbuf.h	2008-02-23 19:44:10.000000000 +0100
@@ -179,6 +179,7 @@
 #define	M_FIRSTFRAG	0x1000	/* packet is first fragment */
 #define	M_LASTFRAG	0x2000	/* packet is last fragment */
 #define	M_VLANTAG	0x10000	/* packet has VLAN tag attached */
+#define	M_IPIN_SKIPPFIL	0x20000 /* m0n0wall: skip pfil processing in ip_input */
 
 /*
  * External buffer types: identify ext_buf type.
