--- sys/boot.orig/i386/boot2/boot1.s	Sat Apr 30 10:14:27 2005
+++ sys/boot/i386/boot2/boot1.s	Sat Apr 30 10:32:38 2005
@@ -195,7 +195,11 @@
 		xorb %al,%al			# Zero assumed bss from
 		rep				#  the end of boot2.bin
 		stosb				#  up to 0x10000
-		callw seta20			# Enable A20
+#		callw seta20			# Enable A20
+		nop
+		nop
+		nop
+		nop
 		jmp start+MEM_JMP-MEM_ORG	# Start BTX
 # 
 # Enable A20 so we can access memory above 1 meg.
