--- sys/geom/part/g_part_bsd.c.orig	2011-03-22 18:43:55.000000000 +0000
+++ sys/geom/part/g_part_bsd.c	2011-03-22 18:45:46.000000000 +0000
@@ -387,11 +387,13 @@
 		goto invalid_label;
 	if (heads != basetable->gpt_heads && !basetable->gpt_fixgeom)
 		basetable->gpt_heads = heads;
+	/*
 	if (sectors != basetable->gpt_sectors || heads != basetable->gpt_heads)
 		printf("GEOM: %s: geometry does not match label"
 		    " (%uh,%us != %uh,%us).\n", pp->name, heads, sectors,
 		    basetable->gpt_heads, basetable->gpt_sectors);
 
+	*/
 	chs = le32dec(buf + 60);
 	if (chs < 1)
 		goto invalid_label;
