--- dhclient-script.orig	Sat Jun 24 15:13:55 2006
+++ dhclient-script	Sat Jun 24 15:28:20 2006
@@ -20,9 +20,7 @@
 #
 
 ARP=/usr/sbin/arp
-AWK=/usr/bin/awk
 HOSTNAME=/bin/hostname
-NETSTAT=/usr/bin/netstat
 
 LOCALHOST=127.0.0.1
 
@@ -135,16 +133,16 @@
 	# thus broke the script. This code creates the resolv.conf if either
 	# are provided.
 
-	local tmpres=/var/run/resolv.conf.${interface}
-	rm -f $tmpres
+	rm -f /var/etc/defaultdomain.conf
+	rm -f /var/etc/nameservers.conf
 
 	if [ -n "$new_domain_name" ]; then
-		echo "search $new_domain_name" >>$tmpres
+		echo $new_domain_name >/var/etc/defaultdomain.conf
 	fi
 
 	if [ -n "$new_domain_name_servers" ]; then
 		for nameserver in $new_domain_name_servers; do
-			echo "nameserver $nameserver" >>$tmpres
+			echo $nameserver >/var/etc/nameservers.conf
 		done
 	fi
 
@@ -208,11 +206,7 @@
 	fi
 fi
 
-if [ -x $NETSTAT ]; then
-	if_defaultroute=`$NETSTAT -rnf inet | $AWK '{if ($1=="default") printf $6}'`
-else
-	if_defaultroute="x"
-fi
+if_defaultroute="x"
 
 case $reason in
 MEDIUM)
@@ -282,14 +276,14 @@
 			fi
 			add_new_routes
 			if add_new_resolv_conf; then
-				exit_with_hooks 0
+				exit 0
 			fi
 		fi
 	fi
 	eval "ifconfig $interface inet -alias $new_ip_address $medium"
 	delete_old_routes
-	exit_with_hooks 1
+	exit 1
 	;;
 esac
 
-exit_with_hooks 0
+exit 0
