diff -ru siproxd-0.6.0/src/log.c siproxd-0.6.0a/src/log.c
--- siproxd-0.6.0/src/log.c	Sat Jun 30 13:25:12 2007
+++ siproxd-0.6.0a/src/log.c	Fri Nov  2 09:45:10 2007
@@ -76,7 +76,7 @@
 static pthread_mutex_t log_mutex = PTHREAD_MUTEX_INITIALIZER;
 
 void log_init(void) {
-   openlog(NULL,LOG_NDELAY|LOG_PID,LOG_DAEMON);
+   openlog(NULL,LOG_NDELAY,LOG_LOCAL6);
 }
 
 void log_end(void) {
@@ -249,7 +249,7 @@
    va_copy(ap_copy, ap);
    vsnprintf(outbuf, sizeof(outbuf), format, ap_copy);
    va_end(ap_copy);
-   syslog(LOG_USER|level, "%s:%i %s%s", file, line, label, outbuf);
+   syslog(LOG_LOCAL6|level, "%s:%i %s%s", file, line, label, outbuf);
    return;
 }
 
diff -ru siproxd-0.6.0/src/utils.c siproxd-0.6.0a/src/utils.c
--- siproxd-0.6.0/src/utils.c	Mon Sep  3 21:55:11 2007
+++ siproxd-0.6.0a/src/utils.c	Fri Nov  2 07:42:21 2007
@@ -23,7 +23,7 @@
 #include <stdlib.h>
 #include <unistd.h>
 #include <errno.h>
-#include <values.h>
+/*#include <values.h>*/
 #include <time.h>
 #include <signal.h>
 #include <string.h>
@@ -33,6 +33,7 @@
 #include <netdb.h>
 #include <net/if.h>
 #include <sys/ioctl.h>
+#include <stdint.h>
 
 #ifdef _SOLARIS2
 # include <sys/sockio.h>
@@ -208,8 +209,8 @@
     */
    j=0;
    k=0;
-   t1=MAXINT;
-   t2=MAXINT;
+   t1=INT32_MAX;
+   t2=INT32_MAX;
    for (i=0; i<DNS_CACHE_SIZE; i++) {
       if (dns_cache[i].hostname[0]=='\0') break;
       if ((dns_cache[i].expires_timestamp < t1) &&
