--- if_re.c.orig	Mon Aug 14 14:43:05 2006
+++ if_re.c	Mon Aug 14 14:42:16 2006
@@ -1235,12 +1235,14 @@
 	ether_ifattach(ifp, eaddr);
 
 	/* Perform hardware diagnostic. */
-	error = re_diag(sc);
+	if (sc->rl_type == RL_8169) {
+		error = re_diag(sc);
 
-	if (error) {
-		device_printf(dev, "attach aborted due to hardware diag failure\n");
-		ether_ifdetach(ifp);
-		goto fail;
+		if (error) {
+			device_printf(dev, "attach aborted due to hardware diag failure\n");
+			ether_ifdetach(ifp);
+			goto fail;
+		}
 	}
 
 	/* Hook interrupt last to avoid having to lock softc */
